core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java

```diff
@@ -127,20 +140,137 @@ public Response deleteBucket(@PathParam("bucket") final String bucket) {
         } catch (Exception e) {
           throw toBucketS3Exception(e, bucketPath);
         }
-        return Response.Status.OK;
+        return Response.Status.NO_CONTENT;
       }
     });
   }
 
   /**
-   * @summary deletes a object
+   * @summary creates an object without using multipart upload
+   * @param contentMD5 the Base64 encoded 128-bit MD5 digest of the object
    * @param bucket the bucket name
    * @param object the object name
+   * @param is the request body
    * @return the response object
    */
+  @PUT
+  @Path(OBJECT_PARAM)
+  @ReturnType("java.lang.Void")
+  @Consumes(MediaType.APPLICATION_OCTET_STREAM)
+  public Response createObject(@HeaderParam("Content-MD5") final String contentMD5,
+                               @PathParam("bucket") final String bucket,
+                               @PathParam("object") final String object,
+                               final InputStream is) {
+    return S3RestUtils.call(bucket, new S3RestUtils.RestCallable<Response>() {
+      @Override
+      public Response call() throws S3Exception {
+        String bucketPath = parseBucketPath(AlluxioURI.SEPARATOR + bucket);
+        checkBucketIsAlluxioDirectory(bucketPath);
+        AlluxioURI objectURI = new AlluxioURI(bucketPath + AlluxioURI.SEPARATOR + object);
+
+        try {
+          CreateFileOptions options = CreateFileOptions.defaults().setRecursive(true)
+              .setWriteType(getS3WriteType());
+          FileOutStream os = mFileSystem.createFile(objectURI, options);
+          MessageDigest md5 = MessageDigest.getInstance("MD5");
+
+          byte[] buf = new byte[BUFFER_SIZE];
+          while (true) {
+            int len = is.read(buf);
+            if (len == -1) {
+              break;
+            }
+            md5.update(buf, 0, len);
+            os.write(buf, 0, len);
```

Comment:
How efficient is this md5 calculation compared to using something like `DigestOutputStream`?